### ISSUE TITLE
Transpose xmap code to mpi4jax

### ISSUE THREAD

**Issue by jecampagne** [mpi4jax/mpi4jax#175]
Hi, 
I would like your help to test a 3D FFT using mpi4jax. I'm currently someting working with `xmap`
```python
from jax.experimental.maps import xmap
from jax.experimental.maps import Mesh


cube_size = 2048
def pnormal(key):
    return jax.random.normal(key, shape=[cube_size])


@partial(xmap,
         in_axes=[...],
         out_axes=['x','y', ...],
         axis_sizes={'x':cube_size, 'y':cube_size},
         axis_resources={'x': 'nx', 'y':'ny',
                         'key_x':'nx', 'key_y':'ny'})

def pnormal(key):
    return jax.random.normal(key, shape=[cube_size])

@partial(xmap,
         in_axes={0:'x', 1:'y'},
         out_axes=['x','y', ...],
         axis_resources={'x': 'nx', 'y': 'ny'})

@jax.jit
def pfft3d(mesh):
    # [x, y, z]
    mesh = jnp.fft.fft(mesh) # Transform on z
    mesh = lax.all_to_all(mesh, 'x', 0, 0) # Now x is exposed, [z,y,x]
    mesh = jnp.fft.fft(mesh) # Transform on x
    mesh = lax.all_to_all(mesh, 'y', 0, 0) # Now y is exposed, [z,x,y]
    mesh = jnp.fft.fft(mesh) # Transform on y
    # [z, x, y]
    return mesh

key = jax.random.PRNGKey(42)
# We reshape all our devices to the mesh shape we want
devices = np.array(jax.devices()).reshape((2, 4))

with Mesh(devices, ('nx', 'ny')):
    mesh = pnormal(key)
    kmesh = pfft3d(mesh)
    kmesh.block_until_ready()
```
So, it is a FFT on each axis of the 3D cube.

Is there a nice way to code it with mpi4jax, eg with 2 machines and 4 GPUs each?
Thanks

**Comment by EiffL** [comment]
I think we can close that issue @jecampagne, I went ahead and implemented all this code with mpi4jax already :-)
https://github.com/DifferentiableUniverseInitiative/JaxPM/blob/mpi/scripts/test_nbody.py

And it works \o/  except when I JIT the code, which is a problem I'll describe in a separate Issue.

**Comment by dionhaefner** [comment]
That's great! Sorry for being slow, I was planning to help but this fell by the wayside. Let me know if you still need help.
